std-bits/table/src/main/java/org/enso/table/data/column/storage/StringStorage.java

```diff
@@ -24,37 +24,39 @@
 /** A column storing strings. */
 public final class StringStorage extends SpecializedStorage<String> {
   private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(StringStorage.class);
-
-  private final TextType type;
   private Future<Long> untrimmedCount;
 
   /**
    * @param data the underlying data
-   * @param size the number of items stored
    * @param type the type of the column
    */
-  public StringStorage(String[] data, int size, TextType type) {
-    super(data, size, buildOps());
-    this.type = type;
+  public StringStorage(String[] data, TextType type) {
+    super(type, data, buildOps());
 
     untrimmedCount =
         CompletableFuture.supplyAsync(
             () -> CountUntrimmed.compute(this, CountUntrimmed.DEFAULT_SAMPLE_SIZE, null));
   }
 
+  public static StringStorage makeEmpty(long size, TextType type) {
+    int intSize = Builder.checkSize(size);
+    return new StringStorage(new String[intSize], type);
+  }
+
   @Override
-  protected SpecializedStorage<String> newInstance(String[] data, int size) {
-    return new StringStorage(data, size, type);
+  public TextType getType() {
+    // As the type is fixed, we can safely cast it.
+    return (TextType) super.getType();
   }
```

Comment:
I'm tempted to make the `Type` also part of the generic arguments of `SpecializedStorage` to avoid these casts.

Please treat as a _very optional suggestion_

Reply:
Tried in 0facc105e1867bd7368f85172a200e591342a48d and actually it is not that simple as best we can do generically is return `StorageType<String>`, but we actually rely on it being `TextType` exactly (as in some places we call specific methods on it).

That's where GADTs would come in and allow us to rely on the _structure_ of `StorageType` being an Algebraic Data Type to refine `StorageType<String>` into only way to construct it - via `TextType`. But Java (unlike Scala) has no GADT support, so that's where we hit the wall. I guess the cast is still needed after all to get the desired behaviour.